plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -79,97 +82,124 @@ Then, in the _body_, we track the term and type application (`Apply` or `TyInst`
+`Utils.CalledVarEnv` with all information needed for inlining saturated functions.
 
 (2) How do we decide whether cost and size are acceptable?
 
 We currently reuse the heuristics 'Utils.sizeIsAcceptable' and 'Utils.costIsAcceptable'
-that are used in 'UnconditionalInline'. For
+that are used in unconditional inlining. For
 
 let v = \x1.\x2...\xn.VBody in body
 
 we check `VBody` with the above "acceptable" functions.
-Note that all `LamAbs` and `TyAbs` should have been
+Note that all `LamAbs` and `TyAbs` are not acceptable in terms of size, but they should have been
 counted out already so we should not immediately encounter those in `VBody`.
 Also, we currently reject `Constant` (has acceptable cost but not acceptable size).
 We may want to check their sizes instead of just rejecting them.
 -}
 
-{-|
-The (syntactic) arity of a term. That is, a record of the arguments that the
-term expects before it may do some work. Since we have both type and lambda
-abstractions, this is not a simple argument count, but rather a list of values
-indicating whether the next argument should be a term or a type.
-
-Note that this is the syntactic arity, i.e. it just corresponds to the number of
-syntactic lambda and type abstractions on the outside of the term. It is thus
-an under-approximation of how many arguments the term may need.
-e.g. consider the term @let id = \x -> x in id@: the variable @id@ has syntactic
-arity @[]@, but does in fact need an argument before it does any work.
--}
-type Arity = [TermOrType]
-
--- | Is the next argument a term or a type?
-data TermOrType =
-    MkTerm | MkType
-    deriving stock (Eq, Show)
-
-instance Pretty TermOrType where
-  pretty = viaShow
-
--- | Computes the 'Arity' of a term.
+-- | Computes the 'Utils.Arity' of a term. Also returns the function body, for checking whether
+-- it's `Utils.acceptable`.
 computeArity ::
-    Term tyname name uni fun a
-    -> Arity
+    Term tyname name uni fun ann
+    -> (Arity, Term tyname name uni fun ann)
 computeArity = \case
-    LamAbs _ _ _ body -> MkTerm : computeArity body
-    TyAbs _ _ _ body  -> MkType : computeArity body
+    LamAbs _ _ _ body -> (first ((:) MkTerm) (computeArity body))
+    TyAbs _ _ _ body  -> (first ((:) MkType) (computeArity body))
     -- Whenever we encounter a body that is not a lambda or type abstraction, we are done counting
-    _                 -> []
+    tm                -> ([],tm)
+
+-- | A term or type argument.
+data Args tyname name uni fun ann =
+  MkTermArg (Term tyname name uni fun ann)
+  | MkTypeArg (Type tyname uni ann)
+
+-- | A list of type or term argument(s) that are being applied.
+type ArgOrder tyname name uni fun ann = [Args tyname name uni fun ann]
+
+-- | A pair of argument and the annotation of the term being applied to,
+-- so a term that was traversed can be built back with `mkApps`.
+type ArgOrderWithAnn tyname name uni fun ann =
```

Comment:
Also the naming confused me again: it's not with _an_ ann, each one has an ann. So following my previous naming suggestions this would be `ArgsWithAnns`.